lib/uplink/encryption.go

```diff
@@ -4,20 +4,179 @@
 package uplink
 
 import (
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/gogo/protobuf/proto"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/encryption"
+	"storj.io/storj/pkg/macaroon"
+	"storj.io/storj/pkg/paths"
+	"storj.io/storj/pkg/pb"
 	"storj.io/storj/pkg/storj"
 )
 
 const (
 	defaultCipher = storj.EncAESGCM
 )
 
-// EncryptionAccess specifies the encryption details needed to encrypt or
-// decrypt objects.
-type EncryptionAccess struct {
-	// Key is the base encryption key to be used for decrypting objects.
-	Key storj.Key
-	// EncryptedPathPrefix is the (possibly empty) encrypted version of the
-	// path from the top of the storage Bucket to this point. This is
-	// necessary to have in order to derive further encryption keys.
-	EncryptedPathPrefix storj.Path
+// EncryptionCtx represents an encryption context. It holds information about
+// how various buckets and objects should be encrypted and decrypted.
+type EncryptionCtx struct {
```

Comment:
This should be probably part of `uplink` not `lib/uplink`.

Reply:
Why? I don't know how to make this decision.

Reply:
The difference is in `lib/uplink` is the higher-level easy-to-use interface, `uplink` is the lower level that allows to have access all the complexity of things. In principle everything should be implementable using only `uplink`, but might be more comfortable with `lib/uplink`.

The `EncryptionCtx` seems more of a lower-level primitive rather than something of higher-level convenience thing.

Reply:
encryption.Store is the lower level primitive. EncryptionCtx is intended to be a higher level convenience thing

Reply:
overall one goal i have is to make lib/uplink usable almost entirely with types defined in lib/uplink for better discoverability when reading the documentation. specifically, doing paginated listing right now is especially brutal to figure out using the documentation with the types not all in the same package

Reply:
@egonelbre can you clarify what the benefit is to the separation of `uplink` and `lib/uplink`? I confess I don't understand.

Reply:
`lib/uplink` is the stable simplified interface for interacting with the network.

`uplink` is the complicated inner workings which allow customization at different levels. e.g. you could switch out how the streaming, encryption etc. work.

Reply:
I _do_ understand the rules you're proposing for what to put in each one, but what I don't understand is _why_. Does it avoid some import cycles or lessen dependencies for a libuplink user? Or is it just that we want some sort of semantic separation between "simple things" and "advanced things" so that users know when they're crossing the line into Hard Mode?

Reply:
More of stable-ish API and unstable-ish API. But also yeah... when you cross over into `uplink` you might lose compatibility with `lib/uplinkc` and so on.

Reply:
due to licensing issues, eventually we need to separate uplink code from storj/storj entirely. when we do that, i would also like to fix the naming. having lib/uplink and uplink is very confusing, independent of the motivation. my plan is that storj.io/uplink will be storj.io/storj/lib/uplink, and storj.io/uplink/internal (or something) will be storj.io/storj/uplink